ping/ping_json.c

```diff
@@ -0,0 +1,241 @@
+// SPDX-License-Identifier: BSD-3-Clause
+/*
+ * Copyright (c) 2024-2025 Georg Pfuetzenreuter <mail+ip@georg-pfuetzenreuter.net>
+ */
+
+#include "ping.h"
+#include <stdarg.h>
+
+static inline void json_emergency(char *msg)
+{
+	error(1, 0, "%s", msg);
+}
+
+static inline void test_buflen(int size)
+{
+	if (size < 0 || size >= PING_JSON_MAX)
+		json_emergency(_("Overflow during JSON construction"));
+}
+
+static void json_end_array(char *part)
+{
+	size_t curlen = strlen(part);
+
+	test_buflen(snprintf(part + curlen, PING_JSON_MAX - curlen, "]"));
+}
+
+static void json_start_object(char *part)
+{
+	test_buflen(snprintf(part, PING_JSON_MAX, "{"));
+}
+
+static void json_end_object(char *part)
+{
+	size_t curlen = strlen(part);
```

Comment:
Wouldn't it make sense to turn the json buffer into a structure that remembers the current offset to the buffer?

Also why do we buffer the data like this, is there a reason why aren't we just printing them into stdout? The only reason I can think of is that an error may happen during a construction of a JSON result data and we do not want to print the partially constructed reply and just emit the error instead.

Reply:
I explained this partially in https://github.com/iputils/iputils/pull/571#issuecomment-2513195980 when different libraries were discussed - your summary is pretty much the same - we do not want to leave behind defective JSON which can break tools processing the piped JSON output.

Reply:
Right, we do not want to print broken JSON. However doing strlen() each time we want to append to the buffer looks ugly to me, I would at least store the offset to the buffer.

Also sorry that I missed comment #571 the ujson library can be trivially made to buffer data when writing, all that is needed is a callback that appends to the buffer instead of the default callback that writes into a file.

Reply:
Sure, that I understand, I will look into storing it in a struct.

And no worries, eventually we concluded not to introduce additional libraries, and I suppose we don't need all the extra functionality ujson provides anytime soon.

Reply:
Adjusted.